### ISSUE TITLE
PKC crash issue with Upnext

### ISSUE THREAD

**Issue by Chalanxe** [croneter/PlexKodiConnect#923]
I am using Coreelec 9.0.3(Stable) for Odroid N2 that has Kodi Leia 18.3. PKC version is  v2.8.7. UpNext version is v1.0.1 (https://forum.kodi.tv/showthread.php?tid=336747) 

I have tried reinstalling both addons on a new sd card with a fresh image but the crash still occurs on each try.(Have only tried with TV Show episodes) This happens at the end of the show after selecting "Watch Now" via UpNext addon. Kodi freezes for about 30 secs to 1min and then crashes.

PKC and Upnext used to play together well about 2 months back and before on Odroid N2.

Thanks for the help!

Debug log:
https://paste.kodi.tv/sugowaqabo.kodi

**Comment by croneter** [comment]
What do the UpNext developers say to this? I can't really see anything in the logs

Workaround for now: use PKC Direct Paths: https://github.com/croneter/PlexKodiConnect/wiki/Set-up-Direct-Paths

**Comment by Chalanxe** [comment]
Will crosspost and see if they can find a bug.Thanks :)

**Comment by Chalanxe** [comment]
This is their reply:
https://github.com/im85288/service.upnext/issues/34#issuecomment-509893691
"It’s not an issue with upnext, the callback to the add-on was sent out. This is a PKC issue with starting playback, I’d post to their issue tracker."

**Comment by angelblue05** [comment]
In your Kodi log, 6051 confirms the callback by Up Next was completed. I see a few PKC lines, right after, but the crash has certainly nothing to do with Up Next.

**Comment by Chalanxe** [comment]
Would the kodi_crashlog.log help?
I only clicked on Watch Now button once,then the current playing show exited back to the episode selection .Any input from remote did not work after that keypress.

**Comment by watchmeexplode5** [comment]
I can also confirm that I'm having this issue as well.  Kodi Leia 18.3. PKC version is v2.8.8 (and also on v2.8.7).
Kodi is on a shield and it crashes kodi when attempting to play next episode.

**Comment by croneter** [comment]
@watchmeexplode5 Are you also using the Kodi add-on UpNext?

**Comment by croneter** [comment]
I can't reproduce this bug on my system. Seems like the Upnext method of starting up a next episode is incompatible with PKC's way of starting playback using add-on paths. 

Does playback work using PKC direct paths?

**Comment by angelblue05** [comment]
@croneter

In up next, the callback method has not changed. What changed is making sure it does not trigger more than one when the next episode starts playing. Up next works correctly for netflix, Embycon, Emby for Kodi, etc. all using add-on playback.

Have you tried on a low powered device? It’s probably the only way to see this happen. It really sounds like a race condition. I’d had a few to deal with over the years with Emby for Kodi, especially with playlist manipulation and starting playback.

The double line I pointed out, was that normal after all?

**Comment by angelblue05** [comment]
Scratch my previous message. UpNext has not changed rhe way playback is initiated (I’ve checked the code again). However, I do see that it’s behaving as if it’s not receiving the play next info from PKC. It’s grabbing the info from Kodi directly and starting playback via jsonrpc. 

@croneter 
Have you tried controlling this by sending your own play next info to upnext? This way you control how playback will initiate? You probably need a slower device to see the current issue. Was that double line I pointed out normal after all? It might be a clue.

Ping me if needed.

**Comment by croneter** [comment]
@angelblue05 thanks for helping out here :wink:. How would I do that and when?

**Comment by angelblue05** [comment]
No problem :)

Here is what you'd need to provide, you want to provide it to upnext only for add-on playback.
Send this data as you prepare to playback or at least 5 minutes before the end of playback. It is up to you, whatever works best for your code.

More detailed: https://github.com/im85288/service.upnext/wiki/Data-to-Send
```
next_info = {
    "current_episode": {
    	"episodeid": string, // Any identifier you want for the episode
    	"tvshowid": string, // Any identifier you want, as long as it is reused for episodes of the same show
    	"title": string, // episode title
    	"showtitle": string, // show title
    	"plot": string, // short plot
    	"playcount": int, // episode playcount
    	"season": string, // season number
    	"episode": string, // episode number
    	"firstaired": string, // year
    	"rating": string,
    	"art": {
    		"tvshow.poster": string, // optional
    		"tvshow.fanart": string, // optional
    		"tvshow.landscape": string, // optional
    		"tvshow.clearart": string, // optional
    		"tvshow.clearlogo": string, // optional
    		"thumb": string // optional
    	}
    },
    "next_episode": {
    	"episodeid": string, // Any identifier you want for the episode
    	"tvshowid": string, // Any identifier you want, as long as it is reused for episodes of the same show
    	"title": string, // episode title
    	"showtitle": string, // show title
    	"plot": string, // short plot
    	"playcount": int, // episode playcount
    	"season": string, // season number
    	"episode": string, // episode number
    	"firstaired": string, // year
    	"rating": string,
    	"art": {
    		"tvshow.poster": string, // optional
    		"tvshow.fanart": string, // optional
    		"tvshow.landscape": string, // optional
    		"tvshow.clearart": string, // optional
    		"tvshow.clearlogo": string, // optional
    		"thumb": string // optional
    	}
    },
    "play_info": {}, // Custom data your add-on provides. Up Next will send it back to initiate playback for the next episode. Fill it with what you need to play the next episode.
    "notification_time": string or int // in seconds. optional
}
```

To send the data. Either use the add-on signals if you know how to work it. Or you can do things a little manually.
```
    import binascii
    import json
    import xbmc

    sender = "plugin.video.plexkodiconnect"
    method = "upnext_data"
    data = '\\"[\\"{0}\\"]\\"'.format(binascii.hexlify(json.dumps(next_info)))

    xbmc.executebuiltin('NotifyAll(%s, %s, %s)' % (sender, method, data))
```

The play_info you give to UpNext will be sent back via xbmc.Monitor onNotification. The sender will start with UPNEXTPROVIDER (I think it may end with .signal), the method will be "plugin.video.plexkodiconnect_play_action".  Your add-on will then be responsible for starting playback as you see fit with the play_info(data) you sent.

**Comment by croneter** [comment]
Does PKC **BETA** version 2.8.11 help?

**Comment by watchmeexplode5** [comment]
2.8.11 has fixed the issue with upnext. Playback works as expected. 
Thank you @croneter and @angelblue05 for resolving the issue so quickly. You  guys are the best!

**Comment by Chalanxe** [comment]
beta 2.8.11 has solved the issue! UpNext works well with PKC now.
Thank you @croneter and @angelblue05 as well for the quick fixes! :)

**Comment by cyberden** [comment]
Hello. It seems that the problem is happening again on Matrix with PKC 3.1.0 and Upnext 1.1.5+matrix.1. on an nvidia shield.

The next episode is notified to upnext but when I click "play", the sound is working but video is blank (white-ish, like if there was an error decoding the video file).

The log is here : https://paste.kodi.tv/paqiqofoyu.kodi but I can't find anything relevant in there....

**Comment by croneter** [comment]
@cyberden  Kodi Matrix still has many bugs. That's one of them. Pkc does not control playback - Kodi does.

**Comment by croneter** [comment]
Should finally be fixed with PlexKodiConnect version 3.6.1
